tools/vms-generator/vms-generator.go

```diff
@@ -171,9 +170,11 @@ func main() {
 		if err != nil {
 			return fmt.Errorf("Failed to open file %v, %v", filename, err)
 		}
-		defer file.Close()
+		defer func() { _ = file.Close() }()
```

Comment:
Ahh... This hurts my soul a bit :smile: 

I really think we're better off without such changes. Basically it adds no value whatsoever and making us write an ugly workaround to comply with the linter.

Can we exclude such `defer` linter errors?

Reply:
I don't know if we can, but, I personally think this is the way to tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not".

I would rather flag these occurrences, and have to jump through the linter's hoop rather than allowing them to remain unchecked in the code base.

Reply:
I think we deal here with two cases: 

 1. we only deal with read-only files. If we can't close them, that is typically not too much of an issue and pretty unlikely. There is probably something pretty wrong on the system, but we can still continue to function in a consistent way. In this case we can ignore the error. Her we could add something like `closeReadOnlyIgnoreErr(fd)`. We may even be able to get from the fd if it is not just read-only and panic, so that we can catch coding errors.
 2. we deal with writeable fds. Here it is more critical that we catch errors and we actually have a bug in our defer if we don't (I expect quite a few). The main background is that flushing may never occur and that while we think that we for instance safed progress in a state file, we may actually never flush the change and don't realize that we never wrote it. This could be for instance an issue if a file system partition gets full.

Reply:
A first naive try from me would go into that direction:

```go
	f, err := os.Open()
	flags, err := unix.FcntlInt(f.Fd(), unix.F_GETFD, 0)
	if flags & (unix.O_WRONLY | unix.O_APPEND | unix.O_RDWR) != 0 {
		panic("this is not a read-only fd")
	}
```

 but that my be incomplete or wrong.

Reply:
Thanks @rmohr . I'll follow up asap.

Reply:
I have to admit I'm not comfortable with it at all.

> I don't know if we can, but, I personally think this is the way to tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not".
>
> I would rather flag these occurrences, and have to jump through the linter's hoop rather than allowing them to remain unchecked in the code base.

After your changes we're being hit twice: we don't check any errors + we tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not". In other words, not you tell the compiler "I don't care about any errors!".

Now how would we able to tell if someone forgot to check the error although it's crucial or explicitly stated that it's fine to not check the error?

IMHO, the bare minimum for this change is to actually handle the crucial errors correctly and distinguish them from the errors we actually don't care about.

Reply:
> I have to admit I'm not comfortable with it at all.
> 
> > I don't know if we can, but, I personally think this is the way to tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not".
> > I would rather flag these occurrences, and have to jump through the linter's hoop rather than allowing them to remain unchecked in the code base.
> 
> After your changes we're being hit twice: we don't check any errors + we tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not". In other words, now you tell the compiler "I don't care about any errors!".
> 
> Now how would we able to tell if someone forgot to check the error although it's crucial or explicitly stated that it's fine to not check the error?

Easy; 

```golang
_ = FuncThatReturnsErr() # implies this is legit. Or whoever writes this is lying. 
FuncThatReturnsErr() # developer forgot to handle the error, is caught by linter.
```

> 
> IMHO, the bare minimum for this change is to actually handle the crucial errors correctly and distinguish them from the errors we actually don't care about.

Meaning this PR's goal is too ambitious for me to tackle right now; this thing caught up tens of errors, and it just looked at the e2e tests ... 

That's OK, we can scope it accordingly, and I'll resume this work once I can spare some time.

Reply:
> > I have to admit I'm not comfortable with it at all.
> > > I don't know if we can, but, I personally think this is the way to tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not".
> > > I would rather flag these occurrences, and have to jump through the linter's hoop rather than allowing them to remain unchecked in the code base.
> > 
> > 
> > After your changes we're being hit twice: we don't check any errors + we tell the compiler "yes, dear compiler, I honestly don't care at all if this errors or not". In other words, now you tell the compiler "I don't care about any errors!".
> > Now how would we able to tell if someone forgot to check the error although it's crucial or explicitly stated that it's fine to not check the error?
> 
> Easy;
> 
> ```go
> _ = FuncThatReturnsErr() // implies this is legit. Or whoever writes this is lying. 
> FuncThatReturnsErr() // developer forgot to handle the error, is caught by linter.
> ```

I'm specifically referring to the `defer FuncThatReturnsErr()` scenario. Sorry if it wasn't clear.
For all other cases I 100% agree.

This is why I would exclude `defer`s from the linter as a start.

> > IMHO, the bare minimum for this change is to actually handle the crucial errors correctly and distinguish them from the errors we actually don't care about.
> 
> Meaning this PR's goal is too ambitious for me to tackle right now; this thing caught up tens of errors, and it just looked at the e2e tests ...
> 
> That's OK, we can scope it accordingly, and I'll resume this work once I can spare some time.